File: AlgorithmsPackCPP/1245. Tree Diameter.cpp
class Solution {
public:
    typedef vector<vector<int>>GRAPH;

    void addEdge(GRAPH&graph, int from, int to){
        graph[from].push_back(to);
        graph[to].push_back(from);
    }

    int mxDiam{};

    int traverse(GRAPH&graph, int node, int parent = -1){
        int h1 = 0;
        int h2 = 0;

        for(int cur:graph[node]){
            if(cur==parent)
                continue;
            
            int value = traverse(graph,cur,node);

            if(value > h1){
                h2 = h1;
                h1 = value;
            }else if(value > h2){
                h2 = value;
            }
        }
        mxDiam = max(mxDiam,h1+h2);
        return 1+h1;
    }

    int treeDiameter(vector<vector<int>>& edges) {
        GRAPH graph(edges.size()+1);
        for(auto&e:edges)
            addEdge(graph,e[0],e[1]);

        traverse(graph,0);
        return mxDiam;
    }
};

File: AlgorithmsPackJavaSolutions/1245. Tree Diameter.java
class Solution {
    private int mxDia = 0;

    private void addEdge(List<List<Integer>> graph, int from, int to) {
        graph.get(from).add(to);
        graph.get(to).add(from);
    }

    private int dfs(List<List<Integer>> graph, int node, int parent) {
        int mxH1 = 0;
        int mxH2 = 0;
        for (int cur : graph.get(node)) {
            if (cur == parent) {
                continue;
            }
            int h = dfs(graph, cur, node);
            if (h > mxH1) {
                mxH2 = mxH1;
                mxH1 = h;
            } else if (h > mxH2) {
                mxH2 = h;
            }
        }
        mxDia = Math.max(mxDia, mxH1 + mxH2);
        return 1 + mxH1;
    }

    public int treeDiameter(int[][] edges) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < edges.length + 1; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] e : edges) {
            addEdge(graph, e[0], e[1]);
        }
        mxDia = 0;
        dfs(graph, 0, -1);
        return mxDia;
    }
}
